atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/knowledge/KnownConcludedTransactionsStoreConcurrencyTest.java

```diff
@@ -0,0 +1,165 @@
+import static org.mockito.ArgumentMatchers.anyMap;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.atMost;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+
+import com.google.common.collect.Range;
+import com.palantir.atlasdb.keyvalue.api.AutoDelegate_KeyValueService;
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.impl.InMemoryKeyValueService;
+import com.palantir.atlasdb.transaction.impl.TransactionConstants;
+import com.palantir.common.concurrent.PTExecutors;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.LongStream;
+import org.junit.Test;
+
+@SuppressWarnings("UnstableApiUsage") // RangeSet usage
+public class KnownConcludedTransactionsStoreConcurrencyTest {
+    private final AtomicBoolean blockCalls = new AtomicBoolean(false);
+    private final CountDownLatch latch = new CountDownLatch(1);
+
+    private final KeyValueService delegateKeyValueService = spy(new InMemoryKeyValueService(true));
+    private final KeyValueService blockingKeyValueService = (AutoDelegate_KeyValueService) () -> {
+        if (blockCalls.get()) {
+            try {
+                latch.await();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException(e);
+            }
+        }
+        return delegateKeyValueService;
+    };
+    private final KnownConcludedTransactionsStore knownConcludedTransactionsStore =
+            KnownConcludedTransactionsStore.create(blockingKeyValueService);
+    private final ExecutorService taskExecutor = PTExecutors.newCachedThreadPool();
+
+    @Test
+    public void batchesReadsUnderHighConcurrency() throws InterruptedException {
+        knownConcludedTransactionsStore.supplement(Range.closedOpen(10L, 50L));
+
+        startBlockingKeyValueServiceCalls();
+        int numThreads = 100;
+
+        List<Future<Optional<TimestampRangeSet>>> readFutures =
+                scheduleTasksInParallel(numThreads, taskIndex -> knownConcludedTransactionsStore.get());
+
+        List<Optional<TimestampRangeSet>> reads = letTasksRunToCompletion(readFutures, false);
+        for (Optional<TimestampRangeSet> read : reads) {
+            assertThat(read).contains(TimestampRangeSet.singleRange(Range.closedOpen(10L, 50L)));
+        }
+
+        verify(delegateKeyValueService, atMost(10))
+                .get(eq(TransactionConstants.KNOWN_CONCLUDED_TRANSACTIONS_TABLE), anyMap());
```

Comment:
Fair enough - in practice this should be `3` almost always (because `supplement(...)` is the 1st call, and then there is a coalescing supplier - the first node to hit it goes first, and then blocks - by the time we release the latch in `letTasksRunToCompletion`, the remaining 99 threads _should_ have reached the coalescing supplier - though of course there is no guarantee that they do).

I'll write a separate branch that runs this test 100 times and we can see how it performs.